pluma/pluma-document.c

```diff
@@ -1422,6 +1451,33 @@ document_loader_loaded (PlumaDocumentLoader *loader,
 		pluma_document_set_newline_type (doc,
 		                                 pluma_document_loader_get_newline_type (loader));
 
+		if (doc->priv->hide_trailing_newline)
+		{
+			gboolean trimmed_trailing_newline;
+			gint doc_char_count;
+
+			g_object_get (loader,
+			              "trimmed-trailing-newline", &trimmed_trailing_newline,
+			              NULL);
+
+			doc_char_count = gtk_text_buffer_get_char_count (GTK_TEXT_BUFFER (doc));
+
+			if (!trimmed_trailing_newline && doc_char_count > 0)
+			{
+				/* Document did not contain any trailing newline, so we want to
+				   change hide-trailing-newline to FALSE so that saving the
+				   document doesn’t automatically add a trailing newline if it
+				   wasn’t previously present. */
+				/* Note that we special-case empty documents here as these never
+				   contain a trailing newline, so we cannot make any assumptions
+				   on whether the omission of the trailing newline was intentional
+				   or not. */
+				g_object_set (doc,
+				              "hide-trailing-newline", FALSE,
+				              NULL);
```

Comment:
This feels wrong. It means that pluma does whatever it wants even when the gsetting is set.